### ISSUE TITLE
Crash on applying any settings

### ISSUE THREAD

**Issue by joonatoona** [libratbag/piper#177]
```
Ubuntu 17.04
Piper from git at commit 4f560b0d1e13383a075dfe94de106942a2028860
libratbag from git at commit 52e10d029f7cc7573f55b63dc435e706c021d72d
Logitech G502 Proteus Spectrum
```

```
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/piper/ratbagd.py", line 192, in callback_handler
    res = obj.call_finish(result)
GLib.Error: g-io-error-quark: Timeout was reached (24)
```

**Comment by Hjdskes** [comment]
The very least we should do here is catch those exceptions so they don't bring Piper down, but ideally we need to figure out why the timeout is occurring in the first place.

I can't find what a timeout means in this asynchronous context. Do you know @whot?
We can pass `GLib.G_MAXINT` to disable it if it doesn't make sense, otherwise we'll have to look into enlarging it as well. For synchronous calls the timeout is 2 seconds, for asynchronous it currently is half a second.

**Comment by muzena** [comment]
@joonatoona 
Can you start piper in Ubuntu 17.04?

**Comment by joonatoona** [comment]
What do you mean? I am in Ubuntu 17.04?

**Comment by muzena** [comment]
@joonatoona 
I have this problem in Ubuntu 17.10:
https://github.com/libratbag/piper/issues/66

**Comment by joonatoona** [comment]
@muzena 

```env PYTHONPATH=$PYTHONPATH:/usr/lib/python3.5/site-packages/ piper```

seems to work just fine

**Comment by muzena** [comment]
Ok, I will try

**Comment by muzena** [comment]
@joonatoona 
I'm trying to compile on clean install Ubuntu 17.04?
I compile and install libratbag
then I try compile ratbag, and tell me that missing libratbag dependencies,
then I install manually libratbag deb packages from   Ubuntu 17.10

compilation gave me this error:
../src/ratbagd-device.c: In function ‘ratbagd_device_get_capabilities’:
../src/ratbagd-device.c:283:3: error: ‘RATBAG_DEVICE_CAP_RESOLUTION’ undeclared (first use in this function)
   RATBAG_DEVICE_CAP_RESOLUTION,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/ratbagd-device.c:283:3: note: each undeclared identifier is reported only once for each function it appears in
../src/ratbagd-device.c:285:3: error: ‘RATBAG_DEVICE_CAP_PROFILE’ undeclared (first use in this function)
   RATBAG_DEVICE_CAP_PROFILE,
   ^~~~~~~~~~~~~~~~~~~~~~~~~
../src/ratbagd-device.c:289:3: error: ‘RATBAG_DEVICE_CAP_BUTTON’ undeclared (first use in this function)
   RATBAG_DEVICE_CAP_BUTTON,
   ^~~~~~~~~~~~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.

Can you in short describe how you compile and install libratbag, ratbag and piper?

**Comment by joonatoona** [comment]
I'm sorry, that's exactly what I did. Maybe one of the maintainers can help you?

**Comment by Hjdskes** [comment]
> Maybe one of the maintainers can help you?

Not in this issue. @muzena Please open a separate issue if you want help, in the libratbag repository as it seems to concern libratbag.

**Comment by muzena** [comment]
@Hjdskes 
Ok nevermind.
Can you tell me, which linux distribution you use or which is the best for piper testing.
I want to check piper translation that I made.

**Comment by joonatoona** [comment]
As they said, this isn't the place to ask.

**Comment by whot** [comment]
ratbagd is part of libratbag now, you should *not* try to build ratbagd anymore. Simply building libratbag is enough.

**Comment by jellopuddingstick** [comment]
I'm experiencing the exact same issue as @joonatoona, Ubuntu 17.04, fresh build/install from master, running with "PYTHONPATH=$PYTHONPATH:/usr/lib/python3.5/site-packages piper", getting a crash whenever I hit "Apply." Let me know if there are any details I can provide to help.

**Comment by Hjdskes** [comment]
I am able to reproduce this with my G403 in wireless mode, so I should be able to debug this in the coming days/weeks.

**Comment by primetoxinz** [comment]
Just got my G502 and this is still happening in my aur build #d243325

**Comment by Hjdskes** [comment]
Sorry, I am side tracked working at my MSc. thesis at the moment. It's on my list of things to look into, but it may be a while.

**Comment by serebit** [comment]
Happens for me on Solus, with python 3.6:
```
Timeout was reached
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/piper/ratbagd.py", line 193, in callback_handler
    res = obj.call_finish(result)
GLib.Error: g-io-error-quark: Timeout was reached (24)
```

**Comment by Hjdskes** [comment]
Can you try again with the latest libratbag and piper from git master? Solus is using our latest release (understandably), which is sadly very old. We should release something new sometime soon.

**Comment by joonatoona** [comment]
I can still reproduce the issue with `piper-git-0.2.900.r30.g5f6ed20` and `libratbag-0.9.902`

**Comment by whot** [comment]
Is this still happening?

**Comment by joonatoona** [comment]
Still happens on 980a214.

**Comment by serebit** [comment]
Fixed for me with the latest release.

**Comment by rainecheck** [comment]
Fresh install from git master of both `piper` and `libratbagd`, getting this error. Ubuntu 18.04.

**Comment by tvannahl** [comment]
Just got my G502 and I can confirm this error.

The error happens on calling the method `Commit`. I just took a short look in the code and found that the following got executed:

Interface: **org.freedesktop.ratbag1.Device**
Object-path: **/org/freedesktop/ratbag1/device/event5**
Method: **Commit**
Value: **()**

I did call that exact method using d-feet 13 times and got the following result:

Average execution time: 6.9730 seconds
Min. execution time: 0.9909 seconds
Max. execution time: 12.9633 seconds.

If I do see that right there are the following issues related to that bug:

 * The timeout of 500ms in `_dbus_call_async` is way to short
 * `_dbus_call_async` appears to be missing the special `TIMED_OUT` handling that is used for example in `_dbus_call`
 * The execution time of `Commit` in ratbagd appears to be quite long.

I am happy to help out in any way I can, but I would love to have some tutoring of how to fix this bug.

**Comment by whot** [comment]
`Commit` should probably be an asynchronous call that always succeeds (barring any obvious errors) and then notifies the caller later with the success status. If you find the time to work on that, that'd be great!
